File: PB - More Exercises JavaScript/6.0 Drawing Figures with Loops - More Exercises/10. Diamond.js
function diamond(n) {
    n = +n
    if (n === 1) return '*'
    if (n === 2) return '**'

    let stars = 1
    if (!(n & 1)) {
        stars = 2
    }

    // up
    for (i = 1; i <= 1; i++) {
        console.log('-'.repeat((Math.ceil(n / 2 - 1))) + '*'.repeat(stars) + '-'.repeat(Math.ceil(n / 2 - 1)))
    }

    //up  /odd
    if (n > 4 && n % 2 !== 0) {
        for (i = 1; i <= (n / 2 - 1); i++) {
            console.log('-'.repeat((n / 2 - i)) + '*' + '-'.repeat((i + i) - 1) + '*' + '-'.repeat(((n / 2 - i))))
        }
    } else {
        for (i = 1; i <= (n / 2 - 2); i++) { //even
            console.log('-'.repeat(((n / 2 - i) - 1)) + '*' + '-'.repeat((i + i)) + '*' + '-'.repeat(((n / 2 - i) - 1)))
        }
    }

    // middle line
    for (i = 0; i < 1; i++) {
        console.log('*' + '-'.repeat((n - 2)) + '*')
    }

    //under middle
    if (n > 4 && n % 2 !== 0) {
        for (i = (n / 2 - 1); i >= 1; i--) {
            console.log('-'.repeat((n / 2 - i)) + '*' + '-'.repeat((i + i) - 2) + '*' + '-'.repeat(((n / 2 - i))))
        }
    } else {
        for (i = (n / 2 - 2); i >= 1; i--) { //even
            console.log('-'.repeat(((n / 2 - i) - 1)) + '*' + '-'.repeat((i + i)) + '*' + '-'.repeat(((n / 2 - i) - 1)))
        }
    }

    // bottom
    for (i = 1; i <= 1; i++) {
        console.log('-'.repeat((Math.ceil(n / 2 - 1))) + '*'.repeat(stars) + '-'.repeat(Math.ceil(n / 2 - 1)))
    }
}

// diamond(3)
// diamond(5)

File: PB - More Exercises GO/6.0 Drawing Figures with Loops - More Exercises/10. Diamond.go
package main

import (
	"fmt"
	"math"
	"os"
	"strings"
)

func main() {
	var n int
	fmt.Scan(&n)
	if n == 1 {
		fmt.Println("*")
		os.Exit(0)
	}

	if n == 2 {
		fmt.Println("**")
		os.Exit(0)
	}

	stars := 1
	if n%2 == 0 {
		stars = 2
	}

	//top row
	for i := 0; i < 1; i++ {
		fix := int(math.Floor(float64(n-1) / 2))
		fmt.Printf("%v%v%v\n", strings.Repeat("-", fix), strings.Repeat("*", stars), strings.Repeat("-", fix))
	}

	//middle top
	if n > 4 && n%2 != 0 { //up  /odd
		for i := 1; i < n/2; i++ {
			fix := int(math.Floor(float64(n/2 - i)))
			midFix := int(math.Floor(float64(i - 1 + i)))
			fmt.Printf("%v*%v*%v\n", strings.Repeat("-", fix), strings.Repeat("-", midFix), strings.Repeat("-", fix))
		}
	} else { //even
		for i := 1; i < n/2-1; i++ {
			fix := int(math.Floor(float64(n/2 - i - 1)))
			fmt.Printf("%v*%v*%v\n", strings.Repeat("-", fix), strings.Repeat("-", i+i), strings.Repeat("-", fix))
		}
	}

	// middle line
	for i := 0; i < 1; i++ {
		fmt.Printf("*%v*\n", strings.Repeat("-", n-2))
	}

	//under middle
	if n > 4 && n%2 != 0 { //odd
		for i := n/2 - 1; i > 0; i-- {
			fix := int(math.Floor(float64(n/2 - i)))
			fmt.Printf("%v*%v*%v\n", strings.Repeat("-", fix), strings.Repeat("-", i+i-1), strings.Repeat("-", fix))
		}

	} else { //even
		for i := n/2 - 1; i > 1; i-- {
			fix := int(math.Floor(float64(n/2 - i)))
			fmt.Printf("%v*%v*%v\n", strings.Repeat("-", fix), strings.Repeat("-", i+i-2), strings.Repeat("-", fix))
		}
	}

	//buttom row
	for i := 0; i < 1; i++ {
		fix := int(math.Floor(float64(n-1) / 2))
		fmt.Printf("%v%v%v\n", strings.Repeat("-", fix), strings.Repeat("*", stars), strings.Repeat("-", fix))
	}

}

/*
name   :10. Diamond
input  :6
output :
--**--
-*--*-
*----*
-*--*-
--**--
*/
